core/sail/fts/lucene-api/src/main/java/org/eclipse/rdf4j/sail/lucene/DistanceQuerySpec.java

```diff
@@ -11,63 +11,100 @@
 
 import org.eclipse.rdf4j.model.IRI;
 import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.Value;
 import org.eclipse.rdf4j.query.algebra.EmptySet;
 import org.eclipse.rdf4j.query.algebra.Extension;
 import org.eclipse.rdf4j.query.algebra.ExtensionElem;
 import org.eclipse.rdf4j.query.algebra.Filter;
+import org.eclipse.rdf4j.query.algebra.FunctionCall;
 import org.eclipse.rdf4j.query.algebra.QueryModelNode;
 import org.eclipse.rdf4j.query.algebra.SingletonSet;
 import org.eclipse.rdf4j.query.algebra.StatementPattern;
+import org.eclipse.rdf4j.query.algebra.ValueConstant;
+import org.eclipse.rdf4j.query.algebra.ValueExpr;
 import org.eclipse.rdf4j.query.algebra.Var;
 
 public class DistanceQuerySpec implements SearchQueryEvaluator {
 
-	private QueryModelNode functionParent;
+	private final FunctionCall distanceFunction;
 
-	private Literal from;
+	private final Literal from;
 
-	private IRI units;
+	private final IRI units;
 
-	private double distance;
+	private final ValueExpr distanceExpr;
 
-	private String distanceVar;
+	private final double distance;
+
+	private final String distanceVar;
 
 	private StatementPattern geoStatement;
 
-	private Filter filter;
+	private String subjectVar;
+
+	private Var contextVar;
+
+	private IRI geoProperty;
+
+	private final String geoVar;
+
+	private final Filter filter;
 
-	public void setFunctionParent(QueryModelNode functionParent) {
-		this.functionParent = functionParent;
+	public DistanceQuerySpec(FunctionCall distanceFunction, ValueExpr distanceExpr, String distVar,
+			Filter filter)
+	{
+		this.distanceFunction = distanceFunction;
+		this.distanceExpr = distanceExpr;
+		this.distanceVar = distVar;
+		this.filter = filter;
+		if (distanceFunction != null) {
+			List<ValueExpr> args = distanceFunction.getArgs();
+			this.from = getLiteral(args.get(0));
+			this.geoVar = getVarName(args.get(1));
+			this.units = getURI(args.get(2));
+		}
+		else {
+			this.from = null;
+			this.geoVar = null;
+			this.units = null;
+		}
+		if (distanceExpr != null) {
+			Literal dist = getLiteral(distanceExpr);
+			this.distance = (dist != null) ? dist.doubleValue() : Double.NaN;
+		}
+		else {
+			this.distance = Double.NaN;
+		}
 	}
 
-	public void setFrom(Literal from) {
+	public DistanceQuerySpec(Literal from, IRI units, double dist, String distVar, IRI geoProperty,
+			String geoVar, String subjectVar, Var contextVar)
+	{
 		this.from = from;
+		this.units = units;
+		this.distance = dist;
+		this.distanceVar = distVar;
+		this.geoProperty = geoProperty;
+		this.geoVar = geoVar;
+		this.subjectVar = subjectVar;
+		this.contextVar = contextVar;
+		this.distanceFunction = null;
+		this.distanceExpr = null;
+		this.filter = null;
 	}
 
 	public Literal getFrom() {
 		return from;
 	}
 
-	public void setUnits(IRI units) {
```

Comment:
I like final instance variables as a pattern, but I agree that this would break the API and doesn't seem necessary for the other changes to be useful.